File: 2021/day02/main.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

auto parse(auto infile)
{
    std::vector<std::pair<std::string, int>> result;
    int value;
    for (std::string command; infile >> command >> value;)
        result.emplace_back(command, value);
    return result;
}

auto part1(auto input)
{
    auto depth = 0ll, horizontal = 0ll;
    for (auto [command, value] : input)
    {
        if (command == "up")
            depth -= value;
        else if (command == "down")
            depth += value;
        else if (command == "forward")
            horizontal += value;
    }
    return horizontal * depth;
}

auto part2(auto input)
{
    auto depth = 0ll, horizontal = 0ll, aim = 0ll;
    for (auto [command, value] : input)
    {
        if (command == "up")
            aim -= value;
        else if (command == "down")
            aim += value;
        else if (command == "forward")
        {
            depth += aim * value;
            horizontal += value;
        }
    }
    return horizontal * depth;
}

int main()
{
    auto input = parse(std::ifstream("2021/day02/data.txt"));
    std::cout << part1(input) << std::endl; // 1840243
    std::cout << part2(input) << std::endl; // 1727785422
    return 0;
}

File: 2021/day02/main.rs
extern crate regex;

use regex::Regex;
use std::fs;

fn parse(path: &str) -> Vec<(String, i64)> {
    let contents = fs::read_to_string(path).expect("Something went wrong reading the file");
    let re = Regex::new(r"(\w+) (\d+)").unwrap();
    return re
        .captures_iter(&contents)
        .map(|cap| (cap[1].to_string(), cap[2].parse::<i64>().unwrap()))
        .collect();
}

fn part1(input: &Vec<(String, i64)>) -> i64 {
    let (depth, horizontal) = input.iter().fold((0i64, 0i64), |a, v| {
        let (depth, horizontal) = a;
        let (command, value) = v;
        if *command == "up" {
            return (depth - value, horizontal);
        } else if *command == "down" {
            return (depth + value, horizontal);
        } else if *command == "forward" {
            return (depth, horizontal + value);
        }
        panic!();
    });
    return horizontal * depth;
}

fn part2(input: &Vec<(String, i64)>) -> i64 {
    let (depth, horizontal, _) = input.iter().fold((0i64, 0i64, 0i64), |a, v| {
        let (depth, horizontal, aim) = a;
        let (command, value) = v;
        if *command == "up" {
            return (depth, horizontal, aim - value);
        } else if *command == "down" {
            return (depth, horizontal, aim + value);
        } else if *command == "forward" {
            return (depth + aim * value, horizontal + value, aim);
        }
        panic!();
    });
    return horizontal * depth;
}

fn main() {
    let input = parse("2021/day02/data.txt");
    println!("{}", part1(&input)); // 1840243
    println!("{}", part2(&input)); // 1727785422
}
